### PR TITLE
all: channel accounts

### PR DESCRIPTION
**Pull request by CaioTeixeira95**
### What

This PR implements channel accounts.

### Why

To have a high rate of transactions.

### Known limitations

Add tests and more channel account commands.

#### PR Structure

- [x] It is not possible to break this PR down into smaller PRs.
- [x] This PR does not mix refactoring changes with feature changes.
- [x] This PR's title starts with name of package that is most changed in the PR, or `all` if the changes are broad or impact many packages.

#### Thoroughness

- [ ] This PR adds tests for the new functionality or fixes.
- [x] All updated queries have been tested (refer to [this](https://stackoverflow.com/a/29163465) check if the data set returned by the updated query is expected to be same as the original one).

#### Release

- [x] This is not a breaking change.
- [x] This is ready to be tested in development.
- [ ] The new functionality is gated with a feature flag if this is not ready for production.

### CODE REVIEW

File: internal/services/account_sponsorship_service.go [hunk 1]
```diff
@@ -33,8 +33,8 @@ func (e ErrOperationNotAllowed) Error() string {
 
 const (
 	// Sufficient to cover three average ledger close time.
-	CreateAccountTxnTimeBounds             = 18
-	CreateAccountTxnTimeBoundsSafetyMargin = 12
+	CreateAccountTxnTimeBounds             = 60
+	CreateAccountTxnTimeBoundsSafetyMargin = 60
```
Comment (marcelosalloum) [inline internal/services/account_sponsorship_service.go:37]: 👋  This change is a bad idea, because the network usually times out in less than 30s. If you use 60 seconds here, you'll add complexity top how you will handle some indeterminate responses.

File: internal/services/account_sponsorship_service.go [hunk 2]
```diff
@@ -113,15 +119,19 @@ func (s *accountSponsorshipService) SponsorAccountCreationTransaction(ctx contex
 		},
 	)
 
-	// TODO: use Channel Accounts instead of using the Distribution Account.
-	distributionAccount, err := s.HorizonClient.AccountDetail(horizonclient.AccountRequest{AccountID: s.SignatureClient.GetDistributionAccountPublicKey()})
+	channelAccountPublicKey, err := s.ChannelAccountSignatureClient.GetAccountPublicKey(ctx)
+	if err != nil {
+		return "", "", fmt.Errorf("getting channel account public key: %w", err)
+	}
```
Comment (marcelosalloum) [inline internal/services/account_sponsorship_service.go:125]: 💡 It'd be a good idea to add a custom error for when there is no account available, or no account at all.

File: internal/data/accounts.go [hunk 3]
```diff
@@ -31,12 +31,19 @@ func (m *AccountModel) Delete(ctx context.Context, address string) error {
...
-	const query = `SELECT EXISTS(SELECT stellar_address FROM accounts WHERE stellar_address = $1)`
+func (m *AccountModel) IsAccountFeeBumpEligible(ctx context.Context, address string) (bool, error) {
+	const query = `
+		SELECT 
+			EXISTS(
+				SELECT stellar_address FROM accounts WHERE stellar_address = $1
+				UNION
+				SELECT public_key FROM channel_accounts WHERE public_key = $1
+			)
+	`
```
Comment (marcelosalloum) [inline internal/data/accounts.go:42]: 👋  I'm confused here... is this method saying that channel accounts can sign a FeeBumpTransaction? Fee bump transactions are used when you have a distribution account paying for the gas fees, so I don't think a Channel Account should be eligible to sign fee bump transactions.

File: internal/services/account_sponsorship_service.go [hunk 4]
```diff
@@ -201,12 +221,16 @@ func (s *accountSponsorshipService) WrapTransaction(ctx context.Context, tx *txn
...
 	}
 
-	return feeBumpTxe, s.SignatureClient.NetworkPassphrase(), nil
+	return feeBumpTxe, s.DistributionAccountSignatureClient.NetworkPassphrase(), nil
 }
 
-func NewAccountSponsorshipService(signatureClient signing.SignatureClient, horizonClient horizonclient.ClientInterface, maxSponsoredBaseReserves int, baseFee int64, models *data.Models) (*accountSponsorshipService, error) {
-	if signatureClient == nil {
-		return nil, fmt.Errorf("signature client cannot be nil")
+func NewAccountSponsorshipService(distributionAccountSignatureClient, channelAccountSignatureClient signing.SignatureClient, horizonClient horizonclient.ClientInterface, maxSponsoredBaseReserves int, baseFee int64, models *data.Models) (*accountSponsorshipService, error) {
```
Comment (marcelosalloum) [inline internal/services/account_sponsorship_service.go:227]: polishing: when we start adding lots of arguments to a method, it starts becoming hard to figure out what they're about. From a caller's perspective, it's hard to grasp what an int vs int64 means unless you take a deeper look into the docs.

I suggest using a struct to pass on these options.

File: cmd/channel_account.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+func (c *channelAccountCmd) Command() *cobra.Command {
+	cfg := channelAccountCmdConfigOptions{}
+	cfgOpts := config.ConfigOptions{
+		utils.DatabaseURLOption(&cfg.DatabaseURL),
+		utils.NetworkPassphraseOption(&cfg.NetworkPassphrase),
+		utils.BaseFeeOption(&cfg.BaseFee),
+		utils.HorizonClientURLOption(&cfg.HorizonClientURL),
+		utils.DistributionAccountPrivateKeyOption(&cfg.DistributionAccountPrivateKey),
```
Comment (marcelosalloum) [inline cmd/channel_account.go:39]: 👋 you should have a separate encryption hey just for encrypting channel accounts. You don't want the host to have to handle the distribution secret when they need to rotate the channel account encryption key.

File: internal/services/channel_account_service.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+	return &channelAccountService{
+		DB:                                 opts.DB,
+		HorizonClient:                      opts.HorizonClient,
+		BaseFee:                            opts.BaseFee,
+		DistributionAccountSignatureClient: opts.DistributionAccountSignatureClient,
+		ChannelAccountStore:                opts.ChannelAccountStore,
+		PrivateKeyEncrypter:                opts.PrivateKeyEncrypter,
+		EncryptionPassphrase:               opts.EncryptionPassphrase,
+	}
+}
```
Comment (marcelosalloum) [inline internal/services/channel_account_service.go:150]: polishing: what's the advantage of such constructor if all you're doing is assign the variables one by one?

File: internal/signing/channelaccounts/channel_accounts_model.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+			FOR UPDATE SKIP LOCKED
+		)
+		RETURNING *
+	`, int64(lockedUntil.Seconds()))
+
+	var channelAccount ChannelAccount
+	for range ChannelAccountWaitTime {
+		err := ca.DB.GetContext(ctx, &channelAccount, query)
+		if errors.Is(err, sql.ErrNoRows) {
+			log.Ctx(ctx).Info("All channel accounts are in use. Retry in 1 second.")
```
Comment (marcelosalloum) [inline internal/signing/channelaccounts/channel_accounts_model.go:68]: 👋 This log should be a warning.

File: internal/signing/channelaccounts/channel_accounts_model.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+				OR locked_until < NOW()
+			ORDER BY random()
+			LIMIT 1
+			FOR UPDATE SKIP LOCKED
+		)
+		RETURNING *
+	`, int64(lockedUntil.Seconds()))
+
+	var channelAccount ChannelAccount
+	for range ChannelAccountWaitTime {
```
Comment (marcelosalloum) [inline internal/signing/channelaccounts/channel_accounts_model.go:65]: 👋 this retry logic should 100% not be in the model. You should remove it and address it in another PR if needed.

File: internal/signing/env_signature_client.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+	"github.com/stellar/go/keypair"
+	"github.com/stellar/go/network"
+	"github.com/stellar/go/txnbuild"
+)
+
+type envSignatureClient struct {
+	networkPassphrase       string
+	distributionAccountFull *keypair.Full
+}
```
Comment (marcelosalloum) [inline internal/signing/env_signature_client.go:15]: Plz, make sure to override the String() method to prevent the secret key from ever being printed. You can do that in another PR.

**Review by marcelosalloum — COMMENTED** [review]
Thanks a lot for putting all of this together.

I've left a few comments below, the most important ones are marked with the emoji 👋

File: internal/services/account_sponsorship_service.go (hunk 1 above)
Comment (CaioTeixeira95) [reply internal/services/account_sponsorship_service.go:37]: I'm sorry, this was only for testing. the correct one is 30 seconds.

File: internal/serve/middleware/middleware.go [hunk 10]
```diff
@@ -47,3 +48,30 @@ func SignatureMiddleware(signatureVerifier auth.SignatureVerifier) func(next htt
 		})
 	}
 }
+
+// RecoverHandler is a middleware that recovers from panics and logs the error.
+func RecoverHandler(next http.Handler) http.Handler {
```
Comment (daniel-burghardt) [inline internal/serve/middleware/middleware.go:53]: Nice! 🚀 
Shall we also add a TODO comment to add sentry logging here?

File: internal/serve/serve.go [hunk 11]
```diff
@@ -140,6 +151,7 @@ func handler(deps handlerDeps) http.Handler {
 			handler := &httphandler.AccountHandler{
 				AccountSponsorshipService: deps.AccountSponsorshipService,
 				SupportedAssets:           deps.SupportedAssets,
+				BlockedOperationsTypes:    blockedOperationTypes,
```
Comment (daniel-burghardt) [inline internal/serve/serve.go:154]: This is unused after the refactor moving it to the service, right?
```suggestion
```

File: internal/services/account_sponsorship_service.go (hunk 2 above)
Comment (daniel-burghardt) [reply internal/services/account_sponsorship_service.go:126]: I think that would be quite useful indeed.

File: internal/services/channel_account_service.go [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	hash, err := signedTx.Hash(s.DistributionAccountSignatureClient.NetworkPassphrase())
+	if err != nil {
+		return fmt.Errorf("getting transaction hash: %w", err)
+	}
+
+	_, err = s.HorizonClient.SubmitTransaction(signedTx)
+	if err != nil {
+		if hError := horizonclient.GetError(err); hError != nil {
```
Comment (daniel-burghardt) [inline internal/services/channel_account_service.go:114]: Maybe we could start thinking of creating a submission service for HTTP requests with retry logic in it (not in this PR for sure). Wdyt of that idea?

File: internal/signing/channel_account_db_signature_client.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+		if _, ok := accountsAlreadyAccountedFor[account]; ok {
+			continue
+		}
+		accountsAlreadyAccountedFor[account] = struct{}{}
+
+		if account == "" {
+			return nil, fmt.Errorf("account %d is empty", i)
+		}
+
+		chAcc, err := sc.channelAccountStore.Get(ctx, sc.dbConnectionPool, account)
```
Comment (daniel-burghardt) [inline internal/signing/channel_account_db_signature_client.go:68]: 💡 if we want to make it more performant, perhaps we could get all accounts from the database at once, before the loop

File: internal/signing/channelaccounts/channel_accounts_model.go [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+		publicKeys[i] = ca.PublicKey
+		encryptedPrivateKeys[i] = ca.EncryptedPrivateKey
+	}
+
+	const q = `
+		INSERT INTO 
+			channel_accounts (public_key, encrypted_private_key)
+		SELECT * 
+			FROM UNNEST($1::text[], $2::text[])
```
Comment (daniel-burghardt) [inline internal/signing/channelaccounts/channel_accounts_model.go:120]: One nice alternative for Postgres is this: https://pkg.go.dev/github.com/lib/pq#hdr-Bulk_imports
But might not be such a good idea because it make the implementation more coupled to one database system.

**Review by daniel-burghardt — COMMENTED** [review]
Great work! 💪 Can't believe you pulled this off in a day 😦.
I'm happy to discuss the comments on a call if you'd like, might be more efficient.

File: internal/signing/channelaccounts/channel_accounts_model.go (hunk 14 above)
Comment (CaioTeixeira95) [reply internal/signing/channelaccounts/channel_accounts_model.go:127]: I prefer keeping it the way it is now.

File: internal/serve/serve.go [hunk 15]
```diff
@@ -126,6 +144,17 @@ func getHandlerDeps(cfg Configs) (handlerDeps, error) {
 	}, nil
 }
 
+func ensureChannelAccounts(channelAccountService services.ChannelAccountService, numberOfChannelAccounts int64) {
+	ctx := context.Background()
+	log.Ctx(ctx).Info("Ensuring the number of channel accounts in the database...")
+	err := channelAccountService.EnsureChannelAccounts(ctx, numberOfChannelAccounts)
+	if err != nil {
+		log.Ctx(ctx).Errorf("error ensuring the number of channel accounts: %s", err.Error())
+		return
```
Comment (daniel-burghardt) [inline internal/serve/serve.go:153]: Whenever we implement Sentry, I think this might be worthy of a Sentry exception as well, wdyt?

File: internal/signing/channel_account_db_signature_client.go [hunk 16]
```diff
@@ -41,18 +43,47 @@ func (sc *channelAccountDBSignatureClient) NetworkPassphrase() string {
...
+	}
+
+	numOfChannelAccounts, err := sc.channelAccountStore.Count(ctx)
 	if err != nil {
-		return "", fmt.Errorf("getting idle channel account: %w", err)
+		return "", fmt.Errorf("getting the number of channel accounts: %w", err)
+	}
+
+	if numOfChannelAccounts > 0 {
+		return "", channelaccounts.ErrNoIdleChannelAccountAvailable
```
Comment (daniel-burghardt) [inline internal/signing/channel_account_db_signature_client.go:66]: 🚀

File: internal/signing/channelaccounts/channel_accounts_model.go [hunk 17]
```diff
@@ -31,6 +31,7 @@ type ChannelAccount struct {
 type ChannelAccountStore interface {
 	GetIdleChannelAccount(ctx context.Context, lockedUntil time.Duration) (*ChannelAccount, error)
 	Get(ctx context.Context, sqlExec db.SQLExecuter, publicKey string) (*ChannelAccount, error)
+	GetAllByPublicKey(ctx context.Context, sqlExec db.SQLExecuter, publicKeys ...string) ([]*ChannelAccount, error)
```
Comment (daniel-burghardt) [inline internal/signing/channelaccounts/channel_accounts_model.go:34]: Do we still need the `Get` function in this case?

File: internal/data/accounts.go (hunk 3 above)
Comment (daniel-burghardt) [reply internal/data/accounts.go:44]: Addressed here: https://github.com/stellar/wallet-backend/blob/049a643b4996aa355e7b6365907d1000fa3f51d4/internal/data/accounts.go#L34-L35

**Review by daniel-burghardt — APPROVED** [review]
LGTM! Awesome! 🚀

File: internal/signing/channelaccounts/channel_accounts_model.go (hunk 17 above)
Comment (CaioTeixeira95) [reply internal/signing/channelaccounts/channel_accounts_model.go:34]: Yes, we do want to preserve compatibility with the TSS, which uses this method.
